### PR TITLE
Performance improvements for dataset load

### PR DESCRIPTION
**Pull request by poornachandra**
@chtyim @abaranau:
I'm going to merge this PR into branch `hotfix/temp-fix-flow-change-instances` to start performance testing. Please still review the changes.

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryFixedModulesDatasetFramework.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  @Override
+  public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagementException, IOException {
+    super.deleteAllInstances(namespaceId);
+  }
+
+  @Override
+  public void addModule(Id.DatasetModule moduleId, DatasetModule module) {
+    throw new UnsupportedOperationException("");
```
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryFixedModulesDatasetFramework.java]: c'mon put some text in it :)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryFixedModulesDatasetFramework.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import javax.annotation.Nullable;
+
+/**
+ * In memory {@link DatasetFramework} implementation that has fixed set of {@link DatasetModule}s.
+ * The modules cannot be changed after the creation of the DatasetFramework.
+ * Since the modules are fixed, this implementation has better performance than {@link InMemoryDatasetFramework}.
+ */
+public class InMemoryFixedModulesDatasetFramework extends InMemoryDatasetFramework implements DatasetFramework {
```
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryFixedModulesDatasetFramework.java]: hate the name.. Hard to figure out good one..

StaticDatasetFramework?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetServiceModules.java [hunk 3]
```diff
@@ -171,4 +173,30 @@ protected void configure() {
       }
     };
   }
+
+  private static final class DatasetMdsProvider implements Provider<DatasetFramework> {
```
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetServiceModules.java]: can we NOT use provider? Seems like we can do bindings without it, no? Making guice module complex doesn't do good

**Comment by abaranau** [comment]
couple minor comments - otherwise looks fine. Hard to follow the whole DatasetFramework hiearchy. We may want to look at refactoring it... Maybe later with work on dataset framework improvements

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetServiceModules.java (hunk 3 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetServiceModules.java]: As discussed offline, we need a provider to combine default modules and dataset meta table util values.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryFixedModulesDatasetFramework.java (hunk 1 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryFixedModulesDatasetFramework.java]: Done
